### DISCUSSION TITLE
Miryoku for Keyball39

### DISCUSSION THREAD

**Discussion by nomm-dot** [manna-harbour/miryoku#371] [Q&A] [answered]
I use Miryoku_ZMK on my corne 5-col and have been extremely happy with it and it was super easy to setup with workflows. I purchased a keyball39 almost a year ago now and have been working on and off trying to get Miryoku_QMK flashed and just end up getting frustrated. I'm confident it's my lack of understanding of QMK and possibly my build environment, but can't seem to wrap my head around it ([309](https://github.com/manna-harbour/miryoku/discussions/309#discussion-6427515)). I'll continue to do work at it, but it'd be great for it to be a supported board, but not sure how much work it'd be considering it's unique features.

**Comment by manna-harbour** [comment, marked as answer]
There are a couple of issues making this more difficult. First, the keyboard definition is not in QMK master (and isn't even in a QMK fork, so can't be merged). Second, Miryoku QMK and other user keymaps were removed from QMK master, see https://github.com/manna-harbour/miryoku/discussions/287. I'm planning on migrating Miryoku QMK to a "userspace" repo, and adding the outboards feature from Miryoku ZMK to Miryoku QMK. That will take care of these issues automatically, but for now building for keyball will require some manual work. 

The first step is to build the default keymap for your keyboard according to https://github.com/Yowkees/keyball/blob/main/qmk_firmware/keyboards/keyball/readme.md.

Next is to copy https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku and https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/layouts/community/split_3x5_3/manna-harbour_miryoku to the corresponding locations in your QMK fork or clone and test it by building Miryoku QMK for any of these keyboards: https://github.com/manna-harbour/miryoku_qmk/tree/miryoku/users/manna-harbour_miryoku#split_3x5_3

The final step is to add a Miryoku mapping for your keyboard. Let me know when you're up to this step and I can make it or give directions if you'd like to do it.

**Reply by nomm-dot** [reply]
Thanks for the detailed reply and for the amazing work you put into this. I've successfully built the firmware at this point, I had to disable RGB to make sure the firmware was not too large, but at this point I cannot figure out how to flash the firmware? I've looked at a few guides that describe moving the .uf2 to the device while it's in bootloader mode, but I've tried that and it did not seem to take effect, and it did not un-mount itself like I've experienced with ZMK on a nice!nano, doubt that matters though. I've also seen posts that say you can't just move the files like that or have to specify which half it is, but I've also seen in the Keyball and Holykeebs(vendor I purchased the board from) docs that for the keyball it's not necessary to specify which side your flashing. All very confusing. I am on Linux using the qmk-cli. Any tips?

**Reply by manna-harbour** [reply]
Unlike ZMK, QMK will build one firmware image that works on both halves on a split, as the firmware will decide which half it's running on at run time, but this may require some configuration or set up. See https://git.sijman.nl/_/qmk_firmware/src/branch/master/docs/feature_split_keyboard.md#setting-handedness.

How you flash depends on the bootloader. If you have a uf2 capable bootloader then you should be able to flash in the same way as ZMK. If not you'll need to use a flashing tool. See https://docs.holykeebs.com/firmware/#flashing. You can ask the vendor or on the QMK discord in the #help channel about that. Just ask about flashing the default firmware because if you ask about Miryoku you'll be send back here!

**Comment by erplsf** [comment]
hi, i've recently had the same issue - i've managed to do it with stock layout like this: (except that my keyball39 doesn't have the ball): https://github.com/Yowkees/keyball/compare/main...erplsf:keyball:main

**Reply by nomm-dot** [reply]
Hey this looks great! I do have a ball and have to disable RGB to get the firmware under the size limit. Unless, `#define RGB_DI_PIN WS2812_DI_PIN` is what keeps it smaller, but I'm guessing your just specifying which pin to control from.

**Reply by erplsf** [reply]
no, that is done within the `rules.mk` file, afaik.
regarding your response to previous post - I've used the "official" miryoku installation guide to first setup the environment, then linked in the keyball folder (from my fork, which has adjustments for miryoku keymap) like the guide in the keyball project says. this allows me to call `qmk flash -c -kb keyball/keyball39 -km manna-harbour_miryoku -e MIRYOKU_ALPHAS=QWERTY -e MIRYOKU_NAV=VI -e MIRYOKU_CLIPBOARD=WIN` to compile and flash the firmware (from under Linux, under Windows, the graphical flashing tool doesn't work)

**Reply by erplsf** [reply]
Feel free to ask more questions if needed, I'm happy to help.
